gapic-common/lib/gapic/common/retry_policy.rb

```diff
@@ -0,0 +1,196 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "gapic/common/error_codes"
+
+module Gapic
+  module Common
+    # Gapic Common retry policy base class.
+    class RetryPolicy
+      DEFAULT_INITIAL_DELAY = 1
+      DEFAULT_MAX_DELAY = 15
+      DEFAULT_MULTIPLIER = 1.3
+      DEFAULT_RETRY_CODES = [].freeze
+      DEFAULT_TIMEOUT = 3600 # One hour
+
+      ##
+      # Create new Gapic::Common:RetryPolicy instance.
+      #
+      # @param initial_delay [Numeric] Initial delay in seconds for this policy.
+      # @param max_delay [Numeric] Maximum delay in seconds for this policy.
+      # @param multiplier [Numeric] The delay scaling factor for each subsequent retry attempt.
+      # @param retry_codes [Array<String|Numeric>] List of retry codes.
+      # @param timeout [Numeric] Timeout threshold value in seconds.
+      #
+      def initialize initial_delay: nil, max_delay: nil, multiplier: nil, retry_codes: nil, timeout: nil
+        instance_variable_set "@initial_delay", initial_delay
+        instance_variable_set "@max_delay", max_delay
+        instance_variable_set "@multiplier", multiplier
+        instance_variable_set "@retry_codes", convert_codes(retry_codes)
+        instance_variable_set "@timeout", timeout
+        instance_variable_set "@retry_block", nil
+        @delay = nil
+      end
+
+      def initial_delay
+        @initial_delay || DEFAULT_INITIAL_DELAY
+      end
+
+      def max_delay
+        @max_delay || DEFAULT_MAX_DELAY
+      end
+
+      def multiplier
+        @multiplier || DEFAULT_MULTIPLIER
+      end
+
+      def retry_codes
+        @retry_codes || DEFAULT_RETRY_CODES
+      end
+
+      def timeout
+        @timeout || DEFAULT_TIMEOUT
+      end
+
+      ##
+      # Customize retry logic based on keyword arguments.
+      #
+      # @param [Block] Custom block yielding whether this policy is retriable based on keyword arguments.
+      def customize_retry &retry_block
```

Comment:
As discusssed, let's not add this until/unless we actually need it. It complicates the design. For the case of this current project, the polling harness already has its own retry logic (included in the block passed to `wait`).

Reply:
I removed all "custom retry" logic and added some minor modifications downstream of `call` or `perform_delay`.

Basically, we used to have`def call` and `def call error`, so I updated accordingly to keep the old API contract.